community/detectors/roxy_wi_cve_2022_31137/src/main/java/com/google/tsunami/plugins/detectors/cves/cve202231137/Cve202231137Detector.java

```diff
@@ -0,0 +1,191 @@
+  private static final String HTTP_PARAMETERS = "alert_consumer=1&ipbackend=\";%s+#";
+
+  private final Clock utcClock;
+  private final HttpClient httpClient;
+  private final int oobSleepDuration;
+
+  private final PayloadGenerator payloadGenerator;
+
+  @Inject
+  Cve202231137Detector(
+      @UtcClock Clock utcClock,
+      HttpClient httpClient,
+      PayloadGenerator payloadGenerator,
+      @OobSleepDuration int oobSleepDuration) {
+
+    this.utcClock = checkNotNull(utcClock);
+    this.httpClient = checkNotNull(httpClient);
+    this.payloadGenerator = checkNotNull(payloadGenerator);
+    this.oobSleepDuration = oobSleepDuration;
+  }
+
+  @Override
+  public DetectionReportList detect(
+      TargetInfo targetInfo, ImmutableList<NetworkService> matchedServices) {
+    logger.atInfo().log("Cve202231137Detector starts detecting.");
+
+    return DetectionReportList.newBuilder()
+        .addAllDetectionReports(
+            matchedServices.stream()
+                .filter(NetworkServiceUtils::isWebService)
+                .filter(this::isRoxyWiWebService)
+                .filter(this::isServiceVulnerable)
+                .map(networkService -> buildDetectionReport(targetInfo, networkService))
+                .collect(toImmutableList()))
+        .build();
+  }
+
+  private boolean isRoxyWiWebService(NetworkService networkService) {
+    // note that this fingerprint phase is only for vulnerable versions,
+    // because newer Roxy-Wi versions have different endpoints for login
+    String targetUrl =
+        NetworkServiceUtils.buildWebApplicationRootUrl(networkService) + "app/login.py";
+    try {
+      Optional<String> response =
+          httpClient.send(get(targetUrl).withEmptyHeaders().build(), networkService).bodyString();
+      return response.isPresent() && response.get().contains("<title>Login page - Roxy-WI</title>");
+    } catch (IOException e) {
+      logger.atWarning().withCause(e).log("Request to target %s failed", networkService);
+    }
+    return false;
+  }
+
+  private boolean isServiceVulnerable(NetworkService networkService) {
+    PayloadGeneratorConfig config =
+        PayloadGeneratorConfig.newBuilder()
+            .setVulnerabilityType(PayloadGeneratorConfig.VulnerabilityType.REFLECTIVE_RCE)
+            .setInterpretationEnvironment(
+                PayloadGeneratorConfig.InterpretationEnvironment.LINUX_SHELL)
+            .setExecutionEnvironment(
+                PayloadGeneratorConfig.ExecutionEnvironment.EXEC_INTERPRETATION_ENVIRONMENT)
+            .build();
+
+    Payload payload = payloadGenerator.generate(config);
+    // Check callback server is enabled
+    if (!payload.getPayloadAttributes().getUsesCallbackServer()) {
+      logger.atInfo().log(
+          "The Tsunami callback server is not setup for this environment, so we cannot confirm the"
+              + " RCE callback");
+      return false;
+    }

```

Comment:
The payload can mutate based on the presence or absence of the callback server, please move the check as a guard of the sleep instruction.

```suggestion
```